admin-dev/themes/new-theme/js/pages/product/components/categories/categories-manager.ts

```diff
@@ -0,0 +1,152 @@
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+import ProductMap from '@pages/product/product-map';
+import ProductEventMap from '@pages/product/product-event-map';
+import CategoryTreeSelector from '@pages/product/components/categories/category-tree-selector';
+import Tags from '@pages/product/components/categories/tags';
+import {EventEmitter} from "events";
+
+const ProductCategoryMap = ProductMap.categories;
+
+export default class CategoriesManager {
+  eventEmitter: EventEmitter;
+
+  categoryTreeSelector: CategoryTreeSelector;
+
+  categoriesContainer: HTMLElement;
+
+  addCategoriesBtn: HTMLElement;
+
+  typeaheadData: Array<any>;
+
+  tags: Tags;
+
+  constructor(eventEmitter: EventEmitter) {
+    this.eventEmitter = eventEmitter;
+    this.categoryTreeSelector = new CategoryTreeSelector(eventEmitter);
+
+    const categoriesContainer = document.querySelector(ProductCategoryMap.categoriesContainer);
+
+    if (!(categoriesContainer instanceof HTMLElement)) {
+      throw '"categoriesContainer" must be a valid HTMLElement';
+    }
+    this.categoriesContainer = categoriesContainer;
+
+    const addCategoriesBtn = this.categoriesContainer.querySelector(ProductCategoryMap.addCategoriesBtn);
+
+    if (!(addCategoriesBtn instanceof HTMLElement)) {
+      throw '"addCategoriesBtn" must be a valid HTMLElement';
+    }
+
+    this.addCategoriesBtn = addCategoriesBtn;
+    this.typeaheadData = [];
+    this.tags = new Tags(
+      eventEmitter,
+      `${ProductCategoryMap.categoriesContainer} ${ProductCategoryMap.tagsContainer}`,
+    );
+    this.tags.render(this.collectCategories());
+    this.renderDefaultCategorySelection();
+    this.listenCategoryChanges();
+    this.listenDefaultCategorySelect();
+    this.initCategoryTreeModal();
+  }
+
+  private initCategoryTreeModal() {
+    this.addCategoriesBtn.addEventListener('click', () => this.categoryTreeSelector.showModal(
+      this.collectCategories(),
+    ));
+    this.eventEmitter.on(ProductEventMap.categories.applyCategoryTreeChanges, (eventData) => {
+      this.tags.render(eventData.categories);
+    });
+  }
+
+  private collectCategories() {
+    // these are at first rendered on page load and later updated dynamically
+    const tagsContainer = this.categoriesContainer.querySelector(ProductCategoryMap.tagsContainer) as HTMLElement;
+    const tags = tagsContainer.querySelectorAll(ProductCategoryMap.tagItem);
+    const categories: { id: number; name: string; isDefault: boolean; }[] = [];
```

Comment:
Honestly, I think these types should be more verbose, it's easier to read like this 
```suggestion
    const categories: Array<{ id: number; name: string; isDefault: boolean; }> = [];
```

Don't you think?

Reply:
If the aim is to improve the readability then we should have an internal type (only scoped for this component for now) `Category` this way we can use `Category[]` as a return type which is much more readable than the full description of the structure

Reply:
yep I agree, just a little problem though, there are 3 category types. 
1. has the `id`, `name` ,`isDefault`;
2. (the one retrieved from api) has `id` , `name`, `active`
3. (I build in javascript from the second one) by adding a breadcrumb.

so idk actually how to handle this, I doubt I can homogenize them :thinking:

Reply:
Don't do optional properties, this would be wrong, I think the best way is to create a common type with common properties, and extend it for these 3 custom types

I think in this case we can declare these 3 types inside the same file, so if we import the file, we get the type too if we wants